src/main/java/net/minestom/server/utils/mojang/MojangUtils.java

```diff
@@ -24,36 +25,108 @@ public final class MojangUtils {
-        return retrieve(String.format(FROM_UUID_URL, uuid));
+        try {
+            return retrieve(String.format(FROM_UUID_URL, uuid));
+        } catch (ServiceNotAvailableException | UsernameDoesNotExistException e) {
+            return null;
+        }
     }
 
+    /**
+     * Gets a {@link JsonObject} with the response from the mojang API
+     * @param username The username as a {@link String}
+     * @return The {@link JsonObject} or {@code null} if the mojang API is down or the username is invalid
+     */
     @Blocking
     public static @Nullable JsonObject fromUsername(@NotNull String username) {
-        return retrieve(String.format(FROM_USERNAME_URL, username));
+        try {
+            return retrieve(String.format(FROM_USERNAME_URL, username));
+        } catch (ServiceNotAvailableException | UsernameDoesNotExistException e) {
+            return null;
+        }
     }
 
-    private static @Nullable JsonObject retrieve(@NotNull String url) {
-        return URL_CACHE.get(url, s -> {
-            try {
-                // Retrieve from the rate-limited Mojang API
-                final String response = URLUtils.getText(url);
-                // If our response is "", that means the url did not get a proper object from the url
-                // So the username or UUID was invalid, and therefore we return null
-                if (response.isEmpty()) {
-                    return null;
-                }
-
-                JsonObject jsonObject = JsonParser.parseString(response).getAsJsonObject();
-                if (jsonObject.has("errorMessage")) {
-                    return null;
-                }
-                return jsonObject;
-            } catch (IOException e) {
-                MinecraftServer.getExceptionManager().handleException(e);
-                throw new RuntimeException(e);
+    /**
+     * Gets the JsonObject from a URL, expects a mojang player URL so the errors might not make sense if it is not
+     * @param url The url to retrieve
+     * @return The {@link JsonObject} of the result
+     * @throws ServiceNotAvailableException When the result is empty
+     * @throws UsernameDoesNotExistException If there is an "errorMessage" field in the JSON
+     */
+    private static @NotNull JsonObject retrieve(@NotNull String url) throws ServiceNotAvailableException, UsernameDoesNotExistException {
+        @Nullable final var cacheResult = URL_CACHE.getIfPresent(url);
+
+        if (cacheResult != null) {
+            return cacheResult;
+        }
+
+        try {
+            // Retrieve from the rate-limited Mojang API
+            final String response = URLUtils.getText(url);
+            // If our response is "", that means the url did not get a proper object from the url
+            // So the username or UUID was invalid, and therefore we return null
+            if (response.isEmpty()) {
+                throw new ServiceNotAvailableException("The Mojang API is down");
```

Comment:
It seems that `URLUtils` already handles this

Reply:
Unless I am misunderstanding something, it returns whatever response body was present with the error (which may be nothing, it may not). It seems a bit unhelpful honestly. Probably another function needs to be added to only return the body if successful or something

Reply:
Maybe a class to wrap the response in that gives the response and a status?